src/core/experimental/memutils.d

```diff
@@ -0,0 +1,277 @@
+/**
+ * Pure D replacement of the C Standard Library basic memory building blocks of string.h
+ * Source: $(DRUNTIMESRC core/experimental/memutils.d)
+ */
+module core.experimental.memutils;
+
+/** memset() implementation */
+
+/**
+ * NOTE(stefanos):
+ * Range-checking is not needed since the user never
+ * pass an `n` (byte count) directly.
+ */
+
+/*
+  If T is an array, set all `dst`'s bytes
+  (whose count is the length of the array times
+  the size of the array element) to `val`.
+  Otherwise, set T.sizeof bytes to `val` starting from the address of `dst`.
+*/
+void memset(T)(ref T dst, const ubyte val)
+{
+    import core.internal.traits : isArray;
+    const uint v = cast(uint) val;
+    static if (isArray!T)
+    {
+        size_t n = dst.length * typeof(dst[0]).sizeof;
+        Dmemset(dst.ptr, v, n);
+    }
+    else
+    {
+        Dmemset(&dst, v, T.sizeof);
+    }
+}
+
+
+version (D_SIMD)
+{
+    version = useSIMD;
+}
+else version (LDC)
+{
+    // LDC always supports SIMD (but doesn't ever set D_SIMD) and
+    // the back-end uses the most appropriate size for every target.
+    version = useSIMD;
+}
+else version (GNU)
+{
+    // GNU does not support SIMD by default. We have to do more complicated
+    // stuff below. So we start by default with useSIMD and decide later.
+    version = useSIMD;
+}
+
+version (useSIMD)
+{
+    /* SIMD implementation
+     */
+    //pragma(msg, "SIMD used");
+    extern(C) private void Dmemset(void *d, const uint val, size_t n)
```

Comment:
Is there any reason for this to be `extern(C)`?

Reply:
No.